### ISSUE TITLE
[v2] root daemon crashes

### ISSUE THREAD

**Issue by an-tex** [telepresenceio/telepresence#1866]
Hi,
I'm observing a consistent root daemon crash upon replaying (potentially big) events from cassandra over the vpn. Here's the daemon.log:

```
2021/07/14 09:59:55.3291 info     : No config found. Using default (from pkg/client/config.go:373)
2021/07/14 09:59:55.3299 info    daemon : --- (from pkg/client/daemon/service.go:147)
2021/07/14 09:59:55.3299 info    daemon : Telepresence daemon v2.3.4 (api v3) starting... (from pkg/client/daemon/service.go:148)
2021/07/14 09:59:55.3299 info    daemon : PID is 70659 (from pkg/client/daemon/service.go:149)
2021/07/14 09:59:55.3300 info    daemon :  (from pkg/client/daemon/service.go:150)
2021/07/14 09:59:55.3302 info    daemon/server-grpc : gRPC server started (from pkg/client/daemon/service.go:207)
2021/07/14 09:59:59.2106 info    daemon/watch-cluster-info : Adding service subnet 10.43.0.0/16 (from pkg/client/daemon/tunrouter.go:256)
2021/07/14 09:59:59.2107 info    daemon/watch-cluster-info : Adding pod subnet 10.42.0.0/24 (from pkg/client/daemon/tunrouter.go:262)
2021/07/14 09:59:59.2114 info    daemon/server-dns : Generated new /etc/resolver/telepresence.local (from pkg/client/daemon/outbound_darwin.go:157)
2021/07/14 09:59:59.2136 info    daemon/server-dns : started command ["killall" "-HUP" "mDNSResponder"] : dexec.pid="70668"
2021/07/14 09:59:59.2138 info    daemon/server-dns :  : dexec.pid="70668" dexec.stream="stdin" dexec.err="EOF"
2021/07/14 09:59:59.2223 info    daemon/server-dns : finished successfully: exit status 0 : dexec.pid="70668"
2021/07/14 09:59:59.2241 info    daemon/server-dns : started command ["killall" "mDNSResponderHelper"] : dexec.pid="70669"
2021/07/14 09:59:59.2244 info    daemon/server-dns :  : dexec.pid="70669" dexec.stream="stdin" dexec.err="EOF"
2021/07/14 09:59:59.2347 info    daemon/server-dns :  : dexec.pid="70669" dexec.stream="stdout+stderr" dexec.data="No matching processes were found\n"
2021/07/14 09:59:59.2348 info    daemon/server-dns : finished with error: exit status 1 : dexec.pid="70669"
2021/07/14 09:59:59.2365 info    daemon/server-dns : started command ["dscacheutil" "-flushcache"] : dexec.pid="70670"
2021/07/14 09:59:59.2366 info    daemon/server-dns :  : dexec.pid="70670" dexec.stream="stdin" dexec.err="EOF"
2021/07/14 09:59:59.2397 info    daemon/server-dns : finished successfully: exit status 0 : dexec.pid="70670"
2021/07/14 09:59:59.2398 info    daemon/server-grpc/conn=3 : setting search paths [...] (from pkg/client/daemon/outbound_darwin.go:164)
2021/07/14 09:59:59.2399 info    daemon/server-grpc/conn=3 : Generated new /etc/resolver/telepresence.kube-public.local (from pkg/client/daemon/outbound_darwin.go:225)
2021/07/14 09:59:59.2401 info    daemon/server-grpc/conn=3 : Generated new /etc/resolver/telepresence.kube-system.local (from pkg/client/daemon/outbound_darwin.go:225)
[...]
2021/07/14 09:59:59.2436 info    daemon/server-grpc/conn=3 : started command ["killall" "-HUP" "mDNSResponder"] : dexec.pid="70671"
2021/07/14 09:59:59.2436 info    daemon/server-grpc/conn=3 :  : dexec.pid="70671" dexec.stream="stdin" dexec.err="EOF"
2021/07/14 09:59:59.2561 info    daemon/server-grpc/conn=3 : finished successfully: exit status 0 : dexec.pid="70671"
2021/07/14 09:59:59.2580 info    daemon/server-grpc/conn=3 : started command ["killall" "mDNSResponderHelper"] : dexec.pid="70672"
2021/07/14 09:59:59.2581 info    daemon/server-grpc/conn=3 :  : dexec.pid="70672" dexec.stream="stdin" dexec.err="EOF"
2021/07/14 09:59:59.2670 info    daemon/server-grpc/conn=3 :  : dexec.pid="70672" dexec.stream="stdout+stderr" dexec.data="No matching processes were found\n"
2021/07/14 09:59:59.2671 info    daemon/server-grpc/conn=3 : finished with error: exit status 1 : dexec.pid="70672"
2021/07/14 09:59:59.2687 info    daemon/server-grpc/conn=3 : started command ["dscacheutil" "-flushcache"] : dexec.pid="70673"
2021/07/14 09:59:59.2689 info    daemon/server-grpc/conn=3 :  : dexec.pid="70673" dexec.stream="stdin" dexec.err="EOF"
2021/07/14 09:59:59.2710 info    daemon/server-grpc/conn=3 : finished successfully: exit status 0 : dexec.pid="70673"
2021/07/14 10:02:46.3004 error   daemon/server-router/TUN reader :    CON tcp 10.42.0.0:53152 -> 10.42.0.37:9042, got RST while idle (from pkg/tun/tcp/handler.go:392)
2021/07/14 10:03:02.1926 error   daemon/server-router/TUN reader :    CON tcp 10.42.0.0:53736 -> 10.42.0.37:9042 failed to write to dispatcher's remote endpoint: EOF (from pkg/tun/tcp/tmgr.go:79)
2021/07/14 10:03:02.1927 error   daemon/server-router/MGR stream : goroutine "/daemon/server-router/MGR stream" exited with error: read from grpc.ClientStream failed: rpc error: code = ResourceExhausted desc = grpc: received message larger than max (7369072 vs. 4194304)
2021/07/14 10:03:02.1928 info    daemon/server-router:shutdown_logger : shutting down (gracefully)...
2021/07/14 10:03:02.1930 info    daemon/server-router:shutdown_status :   final goroutine statuses:
2021/07/14 10:03:02.1930 info    daemon/server-router:shutdown_status :     /daemon/server-router/MGR stream: exited with error
2021/07/14 10:03:02.1930 info    daemon/server-router:shutdown_status :     /daemon/server-router/TUN reader: exited without error
2021/07/14 10:03:02.1930 info    daemon/server-router:shutdown_status :     /daemon/server-router/TUN writer: exited without error
2021/07/14 10:03:02.1931 error   daemon/server-router : goroutine "/daemon/server-router" exited with error: read from grpc.ClientStream failed: rpc error: code = ResourceExhausted desc = grpc: received message larger than max (7369072 vs. 4194304)
2021/07/14 10:03:02.1931 info    daemon/server-dns/Server:shutdown_logger : shutting down (gracefully)...
2021/07/14 10:03:02.1931 info    daemon:shutdown_logger : shutting down (gracefully)...
2021/07/14 10:03:02.1931 info    daemon/server-dns:shutdown_logger : shutting down (gracefully)...
2021/07/14 10:03:02.1984 info    daemon/server-dns : started command ["killall" "-HUP" "mDNSResponder"] : dexec.pid="71608"
2021/07/14 10:03:02.1985 info    daemon/server-dns :  : dexec.pid="71608" dexec.stream="stdin" dexec.err="EOF"
2021/07/14 10:03:02.2061 info    daemon/server-dns : finished successfully: exit status 0 : dexec.pid="71608"
2021/07/14 10:03:02.2098 info    daemon/server-dns : started command ["killall" "mDNSResponderHelper"] : dexec.pid="71609"
2021/07/14 10:03:02.2099 info    daemon/server-dns :  : dexec.pid="71609" dexec.stream="stdin" dexec.err="EOF"
2021/07/14 10:03:02.2167 info    daemon/server-dns :  : dexec.pid="71609" dexec.stream="stdout+stderr" dexec.data="No matching processes were found\n"
2021/07/14 10:03:02.2169 info    daemon/server-dns : finished with error: exit status 1 : dexec.pid="71609"
2021/07/14 10:03:02.2207 info    daemon/server-dns : started command ["dscacheutil" "-flushcache"] : dexec.pid="71610"
2021/07/14 10:03:02.2209 info    daemon/server-dns :  : dexec.pid="71610" dexec.stream="stdin" dexec.err="EOF"
2021/07/14 10:03:02.2214 info    daemon/server-dns : finished successfully: exit status 0 : dexec.pid="71610"
2021/07/14 10:03:02.6938 info    daemon:shutdown_status :   final goroutine statuses:
2021/07/14 10:03:02.6939 info    daemon:shutdown_status :     /daemon/server-dns        : exited without error
2021/07/14 10:03:02.6939 info    daemon:shutdown_status :     /daemon/server-grpc       : exited without error
2021/07/14 10:03:02.6940 info    daemon:shutdown_status :     /daemon/server-router     : exited with error
2021/07/14 10:03:02.6940 info    daemon:shutdown_status :     /daemon/watch-cluster-info: exited without error
2021/07/14 10:03:02.6940 error   daemon : read from grpc.ClientStream failed: rpc error: code = ResourceExhausted desc = grpc: received message larger than max (7369072 vs. 4194304) (from pkg/client/daemon/service.go:227)
telepresence: error: read from grpc.ClientStream failed: rpc error: code = ResourceExhausted desc = grpc: received message larger than max (7369072 vs. 4194304)
```

I guess the last line is the actual issue:
`telepresence: error: read from grpc.ClientStream failed: rpc error: code = ResourceExhausted desc = grpc: received message larger than max (7369072 vs. 4194304)`

K8s Rev: v1.20.8
Telepresence: v2.3.4 (api 3)
Client OS: mac os 11.4

**Comment by an-tex** [comment]
@thallgren this might help https://github.com/kubernetes/kubernetes/pull/63977/files

**Comment by thallgren** [comment]
Thanks for the tip @an-tex . I just put up #1867 to make the size configurable.

**Comment by thallgren** [comment]
Closing since this fix was just released in 2.3.5.

**Comment by an-tex** [comment]
@thallgren 
thanks for handling it so swiftly. I've just updated to 2.3.5 but I'm still seeing the issue. I've tried the following configs in `~/Library/Application Support/telepresence/config.yml`:
```
timeouts:
  trafficManagerAPI: 30s
grpc:
  maxReceiveSize: 10Mi
```
and 
```
timeouts:
  trafficManagerAPI: 30s
grpc:
  maxReceiveSize: 10485760
```
and
```
timeouts:
  trafficManagerAPI: 30s
grpc:
  maxReceiveSize: 10M
```

The config seems to be loaded as when I remove it I get the warning `INFO[0000] No config found. Using default`.

The result is always the same, `daemon.log`:
```
2021/07/16 09:13:22.9847 info    daemon : --- (from pkg/client/daemon/service.go:147)
2021/07/16 09:13:22.9852 info    daemon : Telepresence daemon v2.3.5 (api v3) starting... (from pkg/client/daemon/service.go:148)
2021/07/16 09:13:22.9852 info    daemon : PID is 63518 (from pkg/client/daemon/service.go:149)
2021/07/16 09:13:22.9852 info    daemon :  (from pkg/client/daemon/service.go:150)
2021/07/16 09:13:22.9855 info    daemon/server-grpc : gRPC server started (from pkg/client/daemon/service.go:207)
2021/07/16 09:13:25.8173 info    daemon/watch-cluster-info : Adding service subnet 10.43.0.0/16 (from pkg/client/daemon/tunrouter.go:261)
2021/07/16 09:13:25.8173 info    daemon/watch-cluster-info : Adding pod subnet 10.42.0.0/24 (from pkg/client/daemon/tunrouter.go:267)
2021/07/16 09:13:25.8178 info    daemon/server-dns : Generated new /etc/resolver/telepresence.local (from pkg/client/daemon/outbound_darwin.go:157)
2021/07/16 09:13:25.8200 info    daemon/server-dns : started command ["killall" "-HUP" "mDNSResponder"] : dexec.pid="63522"
2021/07/16 09:13:25.8201 info    daemon/server-dns :  : dexec.pid="63522" dexec.stream="stdin" dexec.err="EOF"
2021/07/16 09:13:25.8339 info    daemon/server-dns : finished successfully: exit status 0 : dexec.pid="63522"
2021/07/16 09:13:25.8367 info    daemon/server-dns : started command ["killall" "mDNSResponderHelper"] : dexec.pid="63523"
2021/07/16 09:13:25.8369 info    daemon/server-dns :  : dexec.pid="63523" dexec.stream="stdin" dexec.err="EOF"
2021/07/16 09:13:25.8498 info    daemon/server-dns :  : dexec.pid="63523" dexec.stream="stdout+stderr" dexec.data="No matching processes were found\n"
2021/07/16 09:13:25.8499 info    daemon/server-dns : finished with error: exit status 1 : dexec.pid="63523"
2021/07/16 09:13:25.8517 info    daemon/server-dns : started command ["dscacheutil" "-flushcache"] : dexec.pid="63524"
2021/07/16 09:13:25.8519 info    daemon/server-dns :  : dexec.pid="63524" dexec.stream="stdin" dexec.err="EOF"
2021/07/16 09:13:25.8546 info    daemon/server-dns : finished successfully: exit status 0 : dexec.pid="63524"
2021/07/16 09:13:25.8547 info    daemon/server-grpc/conn=2 : setting search paths ambassador argocd default kube-node-lease kube-public kube-system minio-operator openebs [...] (from pkg/client/daemon/outbound_darwin.go:164)
2021/07/16 09:13:25.8548 info    daemon/server-grpc/conn=2 : Generated new /etc/resolver/telepresence.reporting-accounts.local (from pkg/client/daemon/outbound_darwin.go:225)
2021/07/16 09:13:25.8550 info    daemon/server-grpc/conn=2 : Generated new /etc/resolver/telepresence.telepresence.local (from pkg/client/daemon/outbound_darwin.go:225)
2021/07/16 09:13:25.8551 info    daemon/server-grpc/conn=2 : Generated new /etc/resolver/telepresence.ambassador.local (from pkg/client/daemon/outbound_darwin.go:225)
2021/07/16 09:13:25.8553 info    daemon/server-grpc/conn=2 : Generated new /etc/resolver/telepresence.default.local (from pkg/client/daemon/outbound_darwin.go:225)
[...]
2021/07/16 09:13:25.8593 info    daemon/server-grpc/conn=2 : started command ["killall" "-HUP" "mDNSResponder"] : dexec.pid="63525"
2021/07/16 09:13:25.8595 info    daemon/server-grpc/conn=2 :  : dexec.pid="63525" dexec.stream="stdin" dexec.err="EOF"
2021/07/16 09:13:25.8784 info    daemon/server-grpc/conn=2 : finished successfully: exit status 0 : dexec.pid="63525"
2021/07/16 09:13:25.8807 info    daemon/server-grpc/conn=2 : started command ["killall" "mDNSResponderHelper"] : dexec.pid="63526"
2021/07/16 09:13:25.8809 info    daemon/server-grpc/conn=2 :  : dexec.pid="63526" dexec.stream="stdin" dexec.err="EOF"
2021/07/16 09:13:25.8945 info    daemon/server-grpc/conn=2 :  : dexec.pid="63526" dexec.stream="stdout+stderr" dexec.data="No matching processes were found\n"
2021/07/16 09:13:25.8947 info    daemon/server-grpc/conn=2 : finished with error: exit status 1 : dexec.pid="63526"
2021/07/16 09:13:25.8966 info    daemon/server-grpc/conn=2 : started command ["dscacheutil" "-flushcache"] : dexec.pid="63527"
2021/07/16 09:13:25.8967 info    daemon/server-grpc/conn=2 :  : dexec.pid="63527" dexec.stream="stdin" dexec.err="EOF"
2021/07/16 09:13:25.8990 info    daemon/server-grpc/conn=2 : finished successfully: exit status 0 : dexec.pid="63527"
2021/07/16 09:15:57.1414 info    daemon/watch-cluster-info : Adding service subnet 10.43.0.0/16 (from pkg/client/daemon/tunrouter.go:261)
2021/07/16 09:15:57.1414 info    daemon/watch-cluster-info : Adding pod subnet 10.42.0.0/24 (from pkg/client/daemon/tunrouter.go:267)
2021/07/16 09:16:00.7480 error   daemon/server-router/MGR stream : goroutine "/daemon/server-router/MGR stream" exited with error: read from grpc.ClientStream failed: rpc error: code = ResourceExhausted desc = grpc: received message larger than max (5592557 vs. 4194304)
2021/07/16 09:16:00.7481 info    daemon/server-router:shutdown_logger : shutting down (gracefully)...
2021/07/16 09:16:00.7483 info    daemon/server-router:shutdown_status :   final goroutine statuses:
2021/07/16 09:16:00.7483 info    daemon/server-router:shutdown_status :     /daemon/server-router/MGR stream: exited with error
2021/07/16 09:16:00.7483 info    daemon/server-router:shutdown_status :     /daemon/server-router/TUN reader: exited without error
2021/07/16 09:16:00.7483 info    daemon/server-router:shutdown_status :     /daemon/server-router/TUN writer: exited without error
2021/07/16 09:16:00.7484 error   daemon/server-router : goroutine "/daemon/server-router" exited with error: read from grpc.ClientStream failed: rpc error: code = ResourceExhausted desc = grpc: received message larger than max (5592557 vs. 4194304)
2021/07/16 09:16:00.7484 info    daemon/server-dns/Server:shutdown_logger : shutting down (gracefully)...
2021/07/16 09:16:00.7484 info    daemon:shutdown_logger : shutting down (gracefully)...
2021/07/16 09:16:00.7484 info    daemon/server-dns:shutdown_logger : shutting down (gracefully)...
2021/07/16 09:16:00.7551 info    daemon/server-dns : started command ["killall" "-HUP" "mDNSResponder"] : dexec.pid="63794"
2021/07/16 09:16:00.7552 info    daemon/server-dns :  : dexec.pid="63794" dexec.stream="stdin" dexec.err="EOF"
2021/07/16 09:16:00.7682 info    daemon/server-dns : finished successfully: exit status 0 : dexec.pid="63794"
2021/07/16 09:16:00.7727 info    daemon/server-dns : started command ["killall" "mDNSResponderHelper"] : dexec.pid="63795"
2021/07/16 09:16:00.7728 info    daemon/server-dns :  : dexec.pid="63795" dexec.stream="stdin" dexec.err="EOF"
2021/07/16 09:16:00.7866 info    daemon/server-dns :  : dexec.pid="63795" dexec.stream="stdout+stderr" dexec.data="No matching processes were found\n"
2021/07/16 09:16:00.7868 info    daemon/server-dns : finished with error: exit status 1 : dexec.pid="63795"
2021/07/16 09:16:00.7908 info    daemon/server-dns : started command ["dscacheutil" "-flushcache"] : dexec.pid="63796"
2021/07/16 09:16:00.7917 info    daemon/server-dns :  : dexec.pid="63796" dexec.stream="stdin" dexec.err="EOF"
2021/07/16 09:16:00.7925 info    daemon/server-dns : finished successfully: exit status 0 : dexec.pid="63796"
2021/07/16 09:16:01.2492 info    daemon:shutdown_status :   final goroutine statuses:
2021/07/16 09:16:01.2493 info    daemon:shutdown_status :     /daemon/server-dns        : exited without error
2021/07/16 09:16:01.2493 info    daemon:shutdown_status :     /daemon/server-grpc       : exited without error
2021/07/16 09:16:01.2493 info    daemon:shutdown_status :     /daemon/server-router     : exited with error
2021/07/16 09:16:01.2493 info    daemon:shutdown_status :     /daemon/watch-cluster-info: exited without error
2021/07/16 09:16:01.2493 error   daemon : read from grpc.ClientStream failed: rpc error: code = ResourceExhausted desc = grpc: received message larger than max (5592557 vs. 4194304) (from pkg/client/daemon/service.go:227)
telepresence: error: read from grpc.ClientStream failed: rpc error: code = ResourceExhausted desc = grpc: received message larger than max (5592557 vs. 4194304)

```

I don't know much about go/grpc but might this setting need to bet set on the [local grpc server](https://github.com/telepresenceio/telepresence/blob/release/v2/pkg/client/daemon/service.go#L216) too? or is it the remote kubernetes api server complaining? on another note, it all worked fine with the legacy telepresence1

**Comment by thallgren** [comment]
@an-tex, that's bad news. I have some difficulties reproducing the problem. Ideally, I'd like to write an automated test that verifies that it all works. Would it be possible for you to make some experiments in your setup?

**Comment by an-tex** [comment]
Yes of course, I'd be happy to help @thallgren . Let me know how. I can certainly build from a dev branch and test it. In the mean time I'll try to reproduce it on a smaller scale. By now I think the issue lies not in receiving big chunks but rather sending them (from the perspective of my local machine) as it seems to happen when bigger snapshots are saved into cassandra running in the remote k8s cluster.

**Comment by an-tex** [comment]
A small update. The issue already occurs on e.g. POSTing >4mb files over REST

**Comment by an-tex** [comment]
Alright here's a small reproducible test case, start an nginx server in your cluster by kube apply:
```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx-test
spec:
  selector:
    matchLabels:
      app: nginx-test
  replicas: 1 
  template:
    metadata:
      labels:
        app: nginx-test
    spec:
      containers:
      - name: nginx
        image: nginx:latest
        ports:
        - containerPort: 80
---
apiVersion: v1
kind: Service
metadata:
  name: nginx-test
spec:
  selector:
    app: nginx-test
  ports:
    - protocol: TCP
      port: 80
      targetPort: 80
```

connect to that cluster using telepresence and run 
`dd if=/dev/zero count=1 bs=1m|http PUT http://nginx-test.default`

this will work as it's only a 1MB body. but trying
`dd if=/dev/zero count=1 bs=8m|http PUT http://nginx-test.default`
with an 8mb body will kill the root daemon with the original error message and this request will timeout.

Note: I've used httpie for the http request as it streams the stdin in comparison to curl which buffers it and sends a Content-Length header which makes nginx reject too big bodys immediately

**Comment by thallgren** [comment]
Huge thanks @an-tex! I'll get right on that first thing next week.

**Comment by an-tex** [comment]
ping @thallgren :)

**Comment by josecv** [comment]
@an-tex Thomas is on holiday 🙂 but I believe #1951 should fix this issue, hopefully we'll be able to include it into the next release

**Comment by thallgren** [comment]
Closed by #1951
